md5sum.cpp

```diff
@@ -0,0 +1,147 @@
+// along with this program; if not, write to the Free Software Foundation,
+// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+//
+// http://savannah.nongnu.org/projects/lmi
+// email: <gchicares@sbcglobal.net>
+// snail: Chicares, 186 Belle Woods Drive, Glastonbury CT 06033, USA
+
+#include "pchfile.hpp"
+
+#include "md5sum.hpp"
+#include "md5.hpp"
+#include "miscellany.hpp"
+#include "assert_lmi.hpp"
+
+#include <fstream>
+
+bool md5sum_vec_from_file
+    (std::string const& filename
+    , std::vector<md5sum_for_file>& result_vec
+    , std::string& error_message
+    )
+{
+    std::ifstream is(filename);
+    if(!is)
+        {
+        error_message = "'";
+        error_message.append(filename);
+        error_message.append("': No such file or directory");
+        return false;
+        }
+
+        std::string line;
+        while(std::getline(is, line))
+            {
+            // The minimal length: 32(md5sum) + 2(spaces) + 1(shortest file name)
+            const size_t minimal_line_length = 35;
+
+            // Ignore comment lines, which begin with a '#' character.
+            if(0 == line.find_first_of('#'))
+                {
+                continue;
+                }
+
+            bool format_error = true;
+            do
+                {
+                const auto space_pos = line.find_first_of(' ');
+                if(std::string::npos == line.find_first_of(' ') ||
+                    line.size() < minimal_line_length)
+                    {
+                    break; // Format error.
+                    }
+
+                    const char second_delimiter = line[space_pos + 1];
+                    if(second_delimiter != ' ' && second_delimiter != '*')
+                        {
+                        break; // Format error.
+                        }
+
+                    std::string md5sum = line.substr(0, space_pos);
+                    std::string file = line.substr(space_pos + 2);
+
+                    if(md5sum.size() != chars_per_formatted_hex_byte * md5len)
+                        {
+                        break; // Format error.
+                        }
+
+                    result_vec.emplace_back(file, md5sum);
+
+                    format_error = false;
+                }
+            while(format_error);
+
+            if(format_error)
+                {
+                error_message = "'";
+                error_message.append(filename);
+                error_message.append("': No properly formatted MD5 checksum lines found");
+                result_vec.clear();
+                return false;
+                }
+            }
+
+    return true;
+}
+
+bool calculate_md5sum
+    (std::string const& filename
+    ,std::string& result_md5sum
+    , std::string& error_message
+    )
+{
+    std::vector<unsigned char> md5;
+    md5.resize(md5len);
+
+    std::FILE* file = std::fopen(filename.c_str(), "rb");
```

Comment:
```suggestion
    std::FILE* const file = std::fopen(filename.c_str(), "rb");
```
Prefer making local variables const if they don't change.